adds expand nodes (macros)

bernt-matthias opened the pull request:
Expand can be used everywhere and adds <expand macro="...">

ExpandIO adds an additional token with the same name as the macro
but in upper case to the commands section. Can only be used as
child of inputs and outputs

bernt-matthias committed:
adds expand nodes (macros)

Expand can be used everywhere and adds <expand macro="...">

ExpandIO adds an additional token with the same name as the macro
but in upper case to the commands section. Can only be used as
child of inputs and outputs

bernt-matthias committed:
add macros

hexylena commented:
Cool! This makes sense. Just going to wait for travis to be back online to merge, if that's OK

bernt-matthias commented:
Not ready yet. I will ping you.

hexylena commented:
Ok, retitled since I assumed it was. Thanks for contributing!

bernt-matthias committed:
subparsers that are a parent are macros

each subparser that is a parent of another is becomes
a two macros (in/out) and a token (commandline).

implemented in MacrosTool

bernt-matthias commented:
I guess this is nearly done. I would suggest that I extent the tests in https://github.com/hexylena/argparse2tool/pull/65 such that you see it in action.

bernt-matthias committed:
doc

bernt-matthias committed:
use macros for undefined stdio, requirements, and tests

if attribute is absent introduce <expand ...> with predefined name

bernt-matthias committed:
improved section handling

- correct identifiers
- create cli for children

bernt-matthias committed:
quote output parameters

bernt-matthias committed:
simplify acceptable_child code

bernt-matthias commented on galaxyxml/tool/parameters/__init__.py:
```diff
@@ -69,14 +73,87 @@ def __init__(
+        super(Macro, self).__init__(**passed_kwargs)
+
+    def acceptable_child(self, child):
+        return issubclass(type(child), XMLParam) \
```
Why `type(child)`?

hexylena replied:
that's a good question. I can't say I have a good answer. Probably just remove it.

bernt-matthias replied:
Actually its correct because `issubclass` requires a class. I mixed this up with `isinstance` which is also used in the `acceptable_child` functions.

bernt-matthias commented on galaxyxml/tool/parameters/__init__.py:
```diff
@@ -69,14 +73,87 @@ def __init__(
+        self.node.text = value
+
+    def acceptable_child(self, child):
+        return issubclass(type(child), XMLParam) \
```
same here `type(child)`?

bernt-matthias commented:
TODOs: 

- [ ] test like example.py

bernt-matthias committed:
add example and test for MacrosTool

- plus docs

- also return at least some TODO in the CLI
  generated for output collections

bernt-matthias committed:
fix linter errors

bernt-matthias committed:
fix mako_identifier bug if None in parents

if there is a when (which has a None identifier)
on the path to the root then the export failed.

also fix linter errors

hexylena commented on examples/tool.xml:
```diff
@@ -8,20 +8,23 @@
 #for $i in $repeat:
---data $data
-#end for]]></command>
+--data $repeat.data
```
is this expected? (haven't looked at the repeat code in a while.)

bernt-matthias replied:
Right, this should be `$i.data`

bernt-matthias committed:
fix access of variables in repeats

hexylena committed:
disable flake until further notice

bernt-matthias committed:
fix flake error

bernt-matthias committed:
Merge branch 'topic/macros' of https://github.com/bernt-matthias/galaxyxml into topic/macros

bernt-matthias committed:
Revert "disable flake until further notice"

This reverts commit 1ebe94ca567dd823170a072c862fe716ba075b33.
